### ISSUE TITLE
Runtime on GPU is not parallel when using jax.config.update("jax_enable_x64", True)

### ISSUE THREAD

**Issue by kx-au** [google/jax#7991]
### Discussed in https://github.com/google/jax/discussions/7990

<div type='discussions-op-text'>

<sup>Originally posted by **kx-au** September 23, 2021</sup>
Hi all,

When I run the code below without `config.update("jax_enable_x64", True)`, i.e. using 32-bit number, I see parallel performance -- running a batch of 1000 `test_fn`  using vmap takes the same time as 1 chain. 

However, when I use 64-bit number, I see an increase in the runtime. The increase is not much, e.g. 1000 chains takes about twice as long as 1 chain. Nevertheless, I do not understand why this is the case. I reproduced the same result in Google Colab, this is the exact printout (this is running the code AFTER the first run, to account for compilation time):

1 parallel batches takes 0.127 seconds.
101 parallel batches takes 0.135 seconds.
201 parallel batches takes 0.147 seconds.
301 parallel batches takes 0.158 seconds.
401 parallel batches takes 0.159 seconds.
501 parallel batches takes 0.171 seconds.
601 parallel batches takes 0.195 seconds.
701 parallel batches takes 0.207 seconds.
801 parallel batches takes 0.212 seconds.
901 parallel batches takes 0.231 seconds.
1001 parallel batches takes 0.242 seconds.

Here is my code.
 
```
from jax.config import config
config.update("jax_enable_x64", True)
config.update('jax_platform_name', 'gpu')
import jax.numpy as jnp
import jax
import time

jax.lib.xla_bridge.get_backend().platform # Make sure it's on GPU

key = jax.random.PRNGKey(420)

@jax.jit
def test_fn(x):
  '''run sigmoid for many times'''
  out = jax.lax.fori_loop(
      lower=1,
      upper=10000,
      body_fun=lambda _, x: jax.nn.sigmoid(x) + 1e-1,
      init_val=x)
  return out

test_fn_parallel = jax.vmap(test_fn, in_axes=(0,))

for n in range(1,1002,100): # number of processes
    subkey, key = jax.random.split(key)
    x = jax.random.normal(subkey,(n,100)) # randoml input
    t1 = time.time()
    out = test_fn_parallel(x)
    t2 = time.time()
    print(f'{n} parallel batches takes {t2-t1:.3f} seconds.')
```

Sidenote: Using only 32-bit number, I also observe the same 'less-than-parallel' performance when I vmap over functions (MCMC chains via lax.scan) that involve large matrix multiplications, neural networks etc. The exact same MCMC implementation yields parallel performance when sampling a 100-dimensional Gaussian, which requires no more than `jnp.sum` and its `jax.grad` gradient. Could complex operations also ruin parallelism?</div>

**Comment by jakevdp** [comment]
Hi - thanks for the question! A couple things:

1. when running micro-benchmarks like this, it is important to use `.block_until_ready()` to account for JAX's [Asynchronous Dispatch](https://jax.readthedocs.io/en/latest/async_dispatch.html). Otherwise, you're likely measuring dispatch time rather than computation time.
2. your question is based on the assumption that `vmap` creates parallelized code: this is incorrect. `vmap` creates vectorized code. On GPU, XLA may execute parts of your program in parallel, but that is unrelated to whether it is expressed using `vmap`. However `fori_loop` will always be executed sequentially.
3. It is expected that GPU runtimes for 64-bit computations will be quite a bit slower than for similar 32-bit computations, with the size of the performance impact dependent on details of the GPU hardware.

Hope that helps!

**Comment by kx-au** [comment]
Hi, thanks for the comprehensive response!

> 1. when running micro-benchmarks like this, it is important to use `.block_until_ready()` to account for JAX's [Asynchronous Dispatch](https://jax.readthedocs.io/en/latest/async_dispatch.html). Otherwise, you're likely measuring dispatch time rather than computation time.

Yes, I've since added .block_until_ready() or print the output etc to make sure the output is computed. The same results persists.

> 2. your question is based on the assumption that `vmap` creates parallelized code: this is incorrect. `vmap` creates vectorized code. On GPU, XLA may execute parts of your program in parallel, but that is unrelated to whether it is expressed using `vmap`. However `fori_loop` will always be executed sequentially.

My bad, I am conflating vectorized code and parallelism. That said, my rough understanding is that vmap vectorizes the code which allows GPU to then take advantage of its vectorized nature to run in parallel. Within each of these individual 'processes', the `fori_loop` remain sequential. Would this be correct? Indeed, whether or not we get O(1) runtime (parallel) depends if the GPU capability and the problem size etc. In the 32-bit number case, as well as other simple experiments, I do observe O(1) runtime when I use vmap over an input dimension of size N, where N varies from 1 to 10^3 or more.

> 3. It is expected that GPU runtimes for 64-bit computations will be quite a bit slower than for similar 32-bit computations, with the size of the performance impact dependent on details of the GPU hardware.

Yes, I've ran a few more tests since and I think this could be the real reason. I've rerun the same experiment at a finer scale, where N varies from 1 to 500 at steps of 10. The different colors are re-runs. X-axis = N, Y-axis = runtime.

![image (4)](https://user-images.githubusercontent.com/35991334/134639265-8dd73ce8-3448-49d3-b493-1325c033b1c1.png)

My best guess is that the GPU did run actually some 200 copies in parallel before it is saturated, then it has to do some housekeeping before it can run the next 200 copies and so on. At the large scale presented in the original post, this manifests as a linear trend.

For 32-bit computations, the runtime in fact also starts to increase beyond some 5000 copies. My inference is that the GPU is saturated in the same way, just that it happens much later for the 32-bit case. (I remain unclear about why the increase is linear and not in steps like the above picture.)
![image (3)](https://user-images.githubusercontent.com/35991334/134635668-96a98568-dd4e-4ba0-a21d-ff5b80838adb.png)

In any case, my best guess is that the observed difference in runtime the 32-bit vs 64-bit code is due to saturation of the GPU.
